kernel: Prune leveldb headers

sedited opened the pull request:
Leveldb headers are currently included in the `dbwrapper.h` file and thus available to many of Bitcoin Core's source files. However, leveldb-specific functionality should be abstracted by the `dbwrapper` and does not need to be available to the rest of the code. Having leveldb included in a widely-used header such as `dbwrapper.h` bloats the entire project's header tree.

The `dbwrapper` is a key component of the libbitcoinkernel library. Future users of this library would not want to contend with  having the leveldb headers exposed and potentially polluting their project's namespace.

For these reasons, the leveldb headers are removed from the `dbwrapper` by moving leveldb-specific code to the implementation file and creating a [pimpl](https://en.cppreference.com/w/cpp/language/pimpl) where leveldb member variables are indispensable. As a final step, the leveldb include flags are removed from the `BITCOIN_INCLUDES` and moved to places where the dbwrapper is compiled.

---

This pull request is part of the [libbitcoinkernel project](https://github.com/bitcoin/bitcoin/issues/27587), and more specifically its stage 1 step 3 "Decouple most non-consensus headers from libbitcoinkernel".

sedited committed:
refactor: Wrap DestroyDB in dbwrapper helper

Wrap leveldb::DestroyDB in a helper function without exposing
leveldb-specifics.

Also, add missing optional include.

The context of this commit is an effort to decouple the dbwrapper header
file from leveldb includes. To this end, the includes are moved to the
dbwrapper implementation file. This is done as part of the kernel
project to reduce the number of required includes for users of the
kernel.

sedited committed:
refactor: Split dbwrapper CDBBatch::Write implementation

Keep the generic serialization in the header, while moving
leveldb-specifics to the implementation file.

The context of this commit is an effort to decouple the dbwrapper header
file from leveldb includes. To this end, the includes are moved to the
dbwrapper implementation file. This is done as part of the kernel
project to reduce the number of required includes for users of the
kernel.

sedited committed:
refactor: Split dbwrapper CDBatch::Erase implementation

Keep the generic serialization in the header, while moving
leveldb-specifics to the implementation file.

The context of this commit is an effort to decouple the dbwrapper header
file from leveldb includes. To this end, the includes are moved to the
dbwrapper implementation file. This is done as part of the kernel
project to reduce the number of required includes for users of the
kernel.

sedited committed:
refactor: Pimpl leveldb::batch for CDBBatch

Hide the leveldb::WriteBatch member variable with a pimpl in order not
to expose it directly in the header.

Also move CDBBatch::Clear to the dbwrapper implementation to use the new
impl_batch.

The context of this commit is an effort to decouple the dbwrapper header
file from leveldb includes. To this end, the includes are moved to the
dbwrapper implementation file. This is done as part of the kernel
project to reduce the number of required includes for users of the
kernel.

sedited committed:
refactor: Split dbwrapper CDBIterator::GetKey implementation

Keep the generic serialization in the header, while moving
leveldb-specifics to the implementation file.

The context of this commit is an effort to decouple the dbwrapper header
file from leveldb includes. To this end, the includes are moved to the
dbwrapper implementation file. This is done as part of the kernel
project to reduce the number of required includes for users of the
kernel.

sedited committed:
refactor: Split dbwrapper CDBIterator::GetValue implementation

Keep the generic serialization in the header, while moving
leveldb-specifics to the implementation file.

The context of this commit is an effort to decouple the dbwrapper header
file from leveldb includes. To this end, the includes are moved to the
dbwrapper implementation file. This is done as part of the kernel
project to reduce the number of required includes for users of the
kernel.

sedited committed:
refactor: Split dbwrapper CDBWrapper::Read implementation

Keep the generic serialization in the header, while moving
leveldb-specifics to the implementation file.

The context of this commit is an effort to decouple the dbwrapper header
file from leveldb includes. To this end, the includes are moved to the
dbwrapper implementation file. This is done as part of the kernel
project to reduce the number of required includes for users of the
kernel.

sedited committed:
refactor: Split dbwrapper CDBWrapper::Exists implementation

Keep the generic serialization in the header, while moving
leveldb-specifics to the implementation file.

The context of this commit is an effort to decouple the dbwrapper header
file from leveldb includes. To this end, the includes are moved to the
dbwrapper implementation file. This is done as part of the kernel
project to reduce the number of required includes for users of the
kernel.

sedited committed:
refactor: Split dbwrapper CDBWrapper::EstimateSize implementation

Keep the generic serialization in the header, while moving
leveldb-specifics to the implementation file.

Since CharCast is no longer needed in the header, move it to the
implementation file.

The context of this commit is an effort to decouple the dbwrapper header
file from leveldb includes. To this end, the includes are moved to the
dbwrapper implementation file. This is done as part of the kernel
project to reduce the number of required includes for users of the
kernel.

sedited committed:
refactor: Move CDBWrapper leveldb members to their own context struct

The context of this commit is an effort to decouple the dbwrapper header
file from leveldb includes. To this end, the includes are moved to the
dbwrapper implementation file. This is done as part of the kernel
project to reduce the number of required includes for users of the
kernel.

sedited committed:
refactor: Split dbwrapper CDBIterator::Seek implementation

Keep the generic serialization in the header, while moving
leveldb-specifics to the implementation file.

The context of this commit is an effort to decouple the dbwrapper header
file from leveldb includes. To this end, the includes are moved to the
dbwrapper implementation file. This is done as part of the kernel
project to reduce the number of required includes for users of the
kernel.

sedited committed:
refactor: Move HandleError to dbwrapper implementation

Make it a static function in dbwrapper.cpp, since it is not used
elsewhere and when left in the header, would expose a leveldb type.

The context of this commit is an effort to decouple the dbwrapper header
file from leveldb includes. To this end, the includes are moved to the
dbwrapper implementation file. This is done as part of the kernel
project to reduce the number of required includes for users of the
kernel.

sedited committed:
refactor: Pimpl leveldb::Iterator for CDBIterator

Hide the leveldb::Iterator member variable with a pimpl in order not to
expose it directly in the header.

Also, move CDBWrapper::NewIterator to the dbwrapper implementation to
use the pimpl for CDBIterator initialziation.

The context of this commit is an effort to decouple the dbwrapper header
file from leveldb includes. To this end, the includes are moved to the
dbwrapper implementation file. This is done as part of the kernel
project to reduce the number of required includes for users of the
kernel.

sedited committed:
refactor: Fix logging.h includes

These were uncovered as missing by the next commit.

sedited committed:
build: Remove leveldb from BITCOIN_INCLUDES

Since leveldb is no longer in our header tree, move its include flags to
whereever dbwrapper.cpp is built.

hebasto commented:
Concept ACK.

dergoegge commented:
Concept ACK

TheCharlatan commented:
Re https://github.com/bitcoin/bitcoin/pull/28186#issuecomment-1657883999

>  I don't think any user should include the dbwrapper either? No user should be directly writing or reading from the db. 

It's hard to speculate on how this will be used in the future. Something calling the db directly for e.g. data science use cases would not be unheard of.

> And if you are worried about the txdb.h include, maybe do a pimpl there? IIRC https://github.com/bitcoin/bitcoin/pull/22242 did that, or something like that.

I am also thinking about the `base.h` include. Though not in the kernel, it pollutes a very significant portion of source files. Implementing a pimpl in `base.h` could work, but I don't think that would really be preferable to this PR here. Getting rid of a big, unused besides a single source file include with a bunch of mostly move-only changes seems like a win to me. If this is not sufficient motivation for you though, I'd be happy to work with a reopened #22242, and separate PR for splitting the `DBOptions` and `DBParams`.

maflcko reviewed (approved):
review ACK 0280dc44d227ae34e8fcbb708833bfe9d66c7b5f 🔨

<details><summary>Show signature</summary>

Signature:

```
untrusted comment: signature from minisign secret key on empty file; verify via: minisign -Vm "${path_to_any_empty_file}" -P RWTRmVTMeKV5noAMqVlsMugDDCyyTSbA3Re5AkUrhvLVln0tSaFWglOw -x "${path_to_this_whole_four_line_signature_blob}"
RUTRmVTMeKV5npGrKx1nqXCw5zeVHdtdYURB/KlyA/LMFgpNCs+SkW9a8N95d+U4AP1RJMi+krxU1A3Yux4bpwZNLvVBKy0wLgM=
trusted comment: review ACK 0280dc44d227ae34e8fcbb708833bfe9d66c7b5f 🔨
EwQOJEQzugff502JSeduHe1I8VbH7Du0bAvETIvbAxyKfRuGdPQJ9HgpXRv4wHXJz97qTnvIhTIdbkKbnucEAg==
```

</details>

Sjors commented on src/dbwrapper.h:
```diff
@@ -67,10 +66,6 @@ class CDBWrapper;
  */
 namespace dbwrapper_private {
 
-/** Handle database error by throwing dbwrapper_error exception.
```
ca6f83d88cb241786676bfb341fc27f6467d39ee: this description, although not amazing, went missing

Sjors reviewed (approved):
utACK 0280dc44d227ae34e8fcbb708833bfe9d66c7b5f modulo 6005165242cf090589934133f01f9dbd496612f1.

I also checked that all the intermediate commits compile and pass the tests (on Ubuntu 23.04).

I'm not familiar with the LevelBD code, but the refactor makes sense to me.

I'll look at 6005165242cf090589934133f01f9dbd496612f1 later since it might get simplified based on @MarcoFalke's suggestion.

TheCharlatan commented:
Thank you for the reviews @Sjors and @MarcoFalke,

Updated 0280dc44d227ae34e8fcbb708833bfe9d66c7b5f -> e8a3a0a0d761c8544e20ffc3ee3d4c1029c44518 ([cleaveLeveldbHeaders_0](https://github.com/TheCharlatan/bitcoin/tree/cleaveLeveldbHeaders_0) -> [cleaveLeveldbHeaders_1](https://github.com/TheCharlatan/bitcoin/tree/cleaveLeveldbHeaders_1),  [compare](https://github.com/TheCharlatan/bitcoin/compare/cleaveLeveldbHeaders_0..cleaveLeveldbHeaders_1))
* Addressed @MarcoFalke's [comment](https://github.com/bitcoin/bitcoin/pull/28186#discussion_r1280179325), added namespace close comment.
* Addressed @MarcoFalke's [comment](https://github.com/bitcoin/bitcoin/pull/28186#discussion_r1280188384), returned comment to its proper place.
* Addressed @MarcoFalke's [comment](https://github.com/bitcoin/bitcoin/pull/28186#discussion_r1280201239), applied `clang-format-diff` suggestion.
* Addressed @MarcoFalke's [comment](https://github.com/bitcoin/bitcoin/pull/28186#discussion_r1280202266), call constructor with curly braces.
* Addressed @Sjors's [comment](https://github.com/bitcoin/bitcoin/pull/28186#discussion_r1280311080), dropped `p` in pimpl naming for both instances.
* Addressed @Sjors's [comment](https://github.com/bitcoin/bitcoin/pull/28186#discussion_r1280383910), re-added comment for handle error function.
* Addressed @MarcoFalke's [comment](https://github.com/bitcoin/bitcoin/pull/28186#discussion_r1280255522), added `DBContext()` helper function. Also polished up the initialization of `m_db_context`.
* Applied @MarcoFalke's [patch](https://github.com/bitcoin/bitcoin/pull/28186#discussion_r1280242366), removed the reader class and replaced it with some simple lambdas.

TheCharlatan commented on src/dbwrapper.h:
```diff
@@ -232,6 +232,36 @@ class CDBWrapper
+        {
+            ssKey << key;
+        }
+    };
```
Thank you, this is much simpler indeed.

Sjors reviewed (approved):
re-ACK e8a3a0a0d761c8544e20ffc3ee3d4c1029c44518

stickies-v commented on src/dbwrapper.h:
```diff
@@ -237,6 +236,7 @@ class CDBWrapper
     bool m_is_memory;
 
     bool ReadImpl(std::function<void(DataStream&)> write_key_to_stream, std::function<void(CDataStream&)> read_value_from_stream) const;
+    bool ExistsImpl(DataStream& ssKey) const;
```
nit: `const DataStream& ssKey`?

stickies-v reviewed (commented):
Concept ACK, did a first run-through of the code but will dig into it deeper.

TheCharlatan commented on src/dbwrapper.h:
```diff
@@ -97,6 +97,8 @@ class CDBBatch
 
     size_t size_estimate{0};
 
+    void WriteImpl(DataStream& ssKey, CDataStream& ssValue);
```
> WriteImpl, ssKey and ssValue are all CDBBatch class members, I'm not sure passing these as parameters makes sense? And if so, I think ssKey should be const.

This gave me pause too, but it felt clearer to me this way. Let me know if you think passing them arguments here is more confusing.

> At first sight, I'm not entirely sure why ssKey and ssValue are CDBBatch members when they get resized and cleared in every operation, so perhaps cleaning that up is the better way to go, but I don't like the ambiguity the current implementation introduces.

AFAIU they don't get resized in every operation. Clear should not be de-allocating its memory and if it is already at the correct capacity, calling `reserve` should not be doing anything.

stickies-v replied:
Alright yeah I was a bit too quick with my initial suggestion, the current implementation makes sense from a (de)alloc perspective, thanks for pointing that both.

> Let me know if you think passing them arguments here is more confusing.

I still don't really like it, but I agree that taking these 2 arguments is the most natural interface for `WriteImpl`. The only reason we have this weirdness is just because `Write` wants to do some allocation optimization. So I'm okay keeping it like this.

Another approach that could work is to make `ssKey` and `ssValue` static vars? It means we would allocate one additional `DataStream` in `Erase`, but I think on the `CDBWrapper` level that's no big deal?

stickies-v replied:
Yeah, unless any objections this would have my preference, less side effects and easier to understand imo:

<details>
<summary>git diff on 26fe348e6e05d9b50614fdcba9cfb7a711ae86dc</summary>

```diff
diff --git a/src/dbwrapper.h b/src/dbwrapper.h
index 54c4685c48..48a6c9af00 100644
--- a/src/dbwrapper.h
+++ b/src/dbwrapper.h
@@ -92,9 +92,6 @@ private:
     const CDBWrapper &parent;
     leveldb::WriteBatch batch;
 
-    DataStream ssKey{};
-    CDataStream ssValue;
-
     size_t size_estimate{0};
 
     void WriteImpl(const DataStream& ssKey, CDataStream& ssValue);
@@ -103,7 +100,7 @@ public:
     /**
      * @param[in] _parent   CDBWrapper that this batch is to be submitted to
      */
-    explicit CDBBatch(const CDBWrapper& _parent) : parent(_parent), ssValue(SER_DISK, CLIENT_VERSION){};
+    explicit CDBBatch(const CDBWrapper& _parent) : parent(_parent){};
 
     void Clear()
     {
@@ -114,6 +111,9 @@ public:
     template <typename K, typename V>
     void Write(const K& key, const V& value)
     {
+        static DataStream ssKey{};
+        static CDataStream ssValue{SER_DISK, CLIENT_VERSION};
+
         ssKey.reserve(DBWRAPPER_PREALLOC_KEY_SIZE);
         ssValue.reserve(DBWRAPPER_PREALLOC_VALUE_SIZE);
         ssKey << key;
@@ -126,6 +126,8 @@ public:
     template <typename K>
     void Erase(const K& key)
     {
+        static DataStream ssKey{};
+
         ssKey.reserve(DBWRAPPER_PREALLOC_KEY_SIZE);
         ssKey << key;
         leveldb::Slice slKey(CharCast(ssKey.data()), ssKey.size());

```
</details>

TheCharlatan replied:
Not sure about making this a static var. Now the allocations survive beyond the lifetime of the `CDBBatch` objects and I am not sure if this is entirely safe.

stickies-v replied:
Good point. Okay, let's mark as resolved.

TheCharlatan commented:
Thank you for the review @stickies-v,

Updated e8a3a0a0d761c8544e20ffc3ee3d4c1029c44518 -> 8c4481ed3713931247e4cedcb5733d3598050eb7 ([cleaveLeveldbHeaders_1](https://github.com/TheCharlatan/bitcoin/tree/cleaveLeveldbHeaders_1) -> [cleaveLeveldbHeaders_2](https://github.com/TheCharlatan/bitcoin/tree/cleaveLeveldbHeaders_2),  [compare](https://github.com/TheCharlatan/bitcoin/compare/cleaveLeveldbHeaders_1..cleaveLeveldbHeaders_2))
* Addressed @stickies-v's [comment](https://github.com/bitcoin/bitcoin/pull/28186#discussion_r1280498519), added `const` qualifier in DestroyDB's `path_str` arg.
* Addressed @stickies-v's [comment](https://github.com/bitcoin/bitcoin/pull/28186#discussion_r1280862595), and [comment](https://github.com/bitcoin/bitcoin/pull/28186#discussion_r1280943436), added `const` qualifiers to the `DataStream` arguments where applicable.

TheCharlatan commented:
Thank you for the re-review @MarcoFalke,

Updated 8c4481ed3713931247e4cedcb5733d3598050eb7 -> 8c4481ed3713931247e4cedcb5733d3598050eb7 ([cleaveLeveldbHeaders_2](https://github.com/TheCharlatan/bitcoin/tree/cleaveLeveldbHeaders_2) -> [cleaveLeveldbHeaders_3](https://github.com/TheCharlatan/bitcoin/tree/cleaveLeveldbHeaders_3),  [compare](https://github.com/TheCharlatan/bitcoin/compare/cleaveLeveldbHeaders_2..cleaveLeveldbHeaders_3))
* Addressed @MarcoFalke's [comment](https://github.com/bitcoin/bitcoin/pull/28186#discussion_r1281421788), moved `CharCast` to implementation file
* Addressed @MarcoFalke's [comment](https://github.com/bitcoin/bitcoin/pull/28186#discussion_r1281421341), using `Span` now instead of `DataStream` types for passing immutable raw bytes.

stickies-v commented on src/dbwrapper.cpp:
It's not immediately obvious to me why we can't stick to our usual pattern of using a `const DataStream& ssKey, CDataStream& ssValue` function signature, and resort to lambdas as is done here. I think the below code is much more straightforward, but perhaps I'm missing some nuance or drawbacks?

<details>
<summary>git diff on f26396732b940095380d76ed77685e0fb11294b8</summary>

```diff
diff --git a/src/dbwrapper.cpp b/src/dbwrapper.cpp
index 1b2548f262..cb27b739fe 100644
--- a/src/dbwrapper.cpp
+++ b/src/dbwrapper.cpp
@@ -335,11 +335,8 @@ std::vector<unsigned char> CDBWrapper::CreateObfuscateKey() const
     return ret;
 }
 
-bool CDBWrapper::ReadImpl(std::function<void(DataStream&)> write_key_to_stream, std::function<void(CDataStream&)> read_value_from_stream) const
+bool CDBWrapper::ReadImpl(const DataStream& ssKey, CDataStream& ssValue) const
 {
-    DataStream ssKey{};
-    ssKey.reserve(DBWRAPPER_PREALLOC_KEY_SIZE);
-    write_key_to_stream(ssKey);
     leveldb::Slice slKey(CharCast(ssKey.data()), ssKey.size());
 
     std::string strValue;
@@ -351,9 +348,8 @@ bool CDBWrapper::ReadImpl(std::function<void(DataStream&)> write_key_to_stream,
         HandleError(status);
     }
     try {
-        CDataStream ssValue{MakeByteSpan(strValue), SER_DISK, CLIENT_VERSION};
+        ssValue << MakeByteSpan(strValue);
         ssValue.Xor(obfuscate_key);
-        read_value_from_stream(ssValue);
     } catch (const std::exception&) {
         return false;
     }
diff --git a/src/dbwrapper.h b/src/dbwrapper.h
index e48d077a80..e9536b8e7a 100644
--- a/src/dbwrapper.h
+++ b/src/dbwrapper.h
@@ -205,7 +205,7 @@ private:
     //! whether or not the database resides in memory
     bool m_is_memory;
 
-    bool ReadImpl(std::function<void(DataStream&)> write_key_to_stream, std::function<void(CDataStream&)> read_value_from_stream) const;
+    bool ReadImpl(const DataStream& ssKey, CDataStream& ssValue) const;
     bool ExistsImpl(const Span<const std::byte>& key) const;
     size_t EstimateSizeImpl(const Span<const std::byte>& key1, const Span<const std::byte>& key2) const;
     auto& DBContext() const { return *Assert(m_db_context); }
@@ -220,9 +220,16 @@ public:
     template <typename K, typename V>
     bool Read(const K& key, V& value) const
     {
-        auto write_key_to_stream{[&key](DataStream& ssKey) { ssKey << key; }};
-        auto read_value_from_stream{[&value](CDataStream& ssValue) { ssValue >> value; }};
-        return ReadImpl(write_key_to_stream, read_value_from_stream);
+        DataStream ssKey{};
+        ssKey.reserve(DBWRAPPER_PREALLOC_KEY_SIZE);
+        ssKey << key;
+
+        CDataStream ssValue{SER_DISK, CLIENT_VERSION};
+        if(ReadImpl(ssKey, ssValue)) {
+            ssValue >> value;
+            return true;
+        }
+        return false;
     }
 
     template <typename K, typename V>

```
</details>

TheCharlatan replied:
I originally did this to keep the stream reading within the try catch block, while changing as few lines in the implementation as possible. We could just return the string value though in the `ReadImpl` and then complete the data reading in the `Read` function.

sedited committed:
refactor: Correct dbwrapper key naming

The ss- prefix should connotate a DataStream variable. Now that these
variables are byte spans, drop the prefix.

stickies-v commented on src/dbwrapper.cpp:
```diff
@@ -127,24 +152,90 @@ static leveldb::Options GetOptions(size_t nCacheSize)
+
 CDBWrapper::CDBWrapper(const DBParams& params)
-    : m_name{fs::PathToString(params.path.stem())}, m_path{params.path}, m_is_memory{params.memory_only}
+    : m_db_context{std::make_unique<LevelDBContext>()}, m_name{fs::PathToString(params.path.stem())}, m_path{params.path}, m_is_memory{params.memory_only}
```
nit: pretty long line

stickies-v commented on src/dbwrapper.cpp:
```diff
@@ -127,24 +152,90 @@ static leveldb::Options GetOptions(size_t nCacheSize)
-    syncoptions.sync = true;
-    options = GetOptions(params.cache_bytes);
-    options.create_if_missing = true;
+    m_db_context = std::make_unique<LevelDBContext>();
```
I think this needs to be removed?

stickies-v commented on src/dbwrapper.h:
```diff
@@ -254,6 +205,11 @@ class CDBWrapper
+    bool ReadImpl(std::function<void(DataStream&)> write_key_to_stream, std::function<void(CDataStream&)> read_value_from_stream) const;
+    bool ExistsImpl(const Span<const std::byte>& key) const;
+    size_t EstimateSizeImpl(const Span<const std::byte>& key1, const Span<const std::byte>& key2) const;
+    auto& DBContext() const { return *Assert(m_db_context); }
```
nit: I think this needs a `LIFETIMEBOUND`.

Also: suggested alternative name `DBContextRef()` to highlight that the return value can mutate our `CDBWrapper` state? No strong view on it though, happy with the current one too, just a suggestion.

stickies-v commented on src/dbwrapper.cpp:
```diff
@@ -127,24 +152,90 @@ static leveldb::Options GetOptions(size_t nCacheSize)
-    options = GetOptions(params.cache_bytes);
-    options.create_if_missing = true;
+    m_db_context = std::make_unique<LevelDBContext>();
+    DBContext().penv = nullptr;
```
nit: Would this be an improvement, to highlight the reference nature of `db_context`, as well as I think generally it's slightly cleaner to not call `DBContext()` every time (I know it's a very cheap call)?

```suggestion
    auto& db_context{DBContext()};
    db_context.penv = nullptr;
```

TheCharlatan commented:
Updated f26396732b940095380d76ed77685e0fb11294b8 -> 3fb2dac2ce78092c1006ee082c536bea1b69a152 ([cleaveLeveldbHeaders_3](https://github.com/TheCharlatan/bitcoin/tree/cleaveLeveldbHeaders_3) -> [cleaveLeveldbHeaders_4](https://github.com/TheCharlatan/bitcoin/tree/cleaveLeveldbHeaders_4),  [compare](https://github.com/TheCharlatan/bitcoin/compare/cleaveLeveldbHeaders_3..cleaveLeveldbHeaders_4))
* Addressed @MarcoFalke's [comment](https://github.com/bitcoin/bitcoin/pull/28186#discussion_r1283223804), changed `const Span<const std::byte>&` to `Span<const std::byte>`.
* Split out renaming of some of the key variables into a separate commit at the end.

stickies-v commented on src/dbwrapper.cpp:
```diff
@@ -185,16 +276,16 @@ CDBWrapper::CDBWrapper(const DBParams& params)
-    options.block_cache = nullptr;
-    delete penv;
-    options.env = nullptr;
+    delete DBContext().pdb;
+    DBContext().pdb = nullptr;
+    delete DBContext().options.filter_policy;
+    DBContext().options.filter_policy = nullptr;
+    delete DBContext().options.info_log;
+    DBContext().options.info_log = nullptr;
+    delete DBContext().options.block_cache;
+    DBContext().options.block_cache = nullptr;
+    delete DBContext().penv;
+    DBContext().options.env = nullptr;
```
nit: would it make sense to move this to the `LevelDBContext` destructor?

TheCharlatan replied:
It would, but then we'd also have to forward-declare a destructor class for it and I don't think it's really worth doing that.

stickies-v replied:
Fair enough. Can easily improve later if/when necessary.

stickies-v reviewed (commented):
light crACK 3fb2dac2ce78092c1006ee082c536bea1b69a152

stickies-v commented on src/dbwrapper.h:
```diff
@@ -209,30 +177,13 @@ class CDBIterator
-
-    //! the database itself
-    leveldb::DB* pdb;
+    std::unique_ptr<LevelDBContext> m_db_context;
```
nit: missing docstring

stickies-v reviewed (approved):
ACK 3fb2dac2ce78092c1006ee082c536bea1b69a152

I would however strongly prefer to [simplify the lambda logic](https://github.com/bitcoin/bitcoin/pull/28186#discussion_r1283260447), and probably [this duplication should also be removed](https://github.com/bitcoin/bitcoin/pull/28186#discussion_r1283281451).

TheCharlatan commented:
Thank you for the review @stickies-v,

Updated 3fb2dac2ce78092c1006ee082c536bea1b69a152 -> 1be04724a3ac061859118c09b90e2e15ea8d04b0 ([cleaveLeveldbHeaders_4](https://github.com/TheCharlatan/bitcoin/tree/cleaveLeveldbHeaders_4) -> [cleaveLeveldbHeaders_5](https://github.com/TheCharlatan/bitcoin/tree/cleaveLeveldbHeaders_5),  [compare](https://github.com/TheCharlatan/bitcoin/compare/cleaveLeveldbHeaders_4..cleaveLeveldbHeaders_5))

* Addressed @stickies-v's [comment_1](https://github.com/bitcoin/bitcoin/pull/28186#discussion_r1283098892) and [comment_2](https://github.com/bitcoin/bitcoin/pull/28186#discussion_r1283280722), added some line breaks in constructor lists.
* Addressed @stickies-v's [comment](https://github.com/bitcoin/bitcoin/pull/28186#discussion_r1283260447), further split up the `CDBWrapper::Read` implementation. The `ReadImpl` now returns an option of the read database value string.
* Addressed @stickies-v's [comment](https://github.com/bitcoin/bitcoin/pull/28186#discussion_r1283281451), removed left over `m_db_context` initializer.
* Addressed @stickies-v's [comment](https://github.com/bitcoin/bitcoin/pull/28186#discussion_r1283286314), added `LIFETIMEBOUND` to `DBContext` reference getter.
* Addressed @stickies-v's [comment](https://github.com/bitcoin/bitcoin/pull/28186#discussion_r1284419848), added docstring to `m_db_contex` `CDBWrapper` member variable.

stickies-v commented on src/dbwrapper.h:
```diff
@@ -5,28 +5,22 @@
-#include <leveldb/slice.h>
-#include <leveldb/status.h>
-#include <leveldb/write_batch.h>
+#include <functional>
```
nit: No longer needed, I think? (in 9a63566be5445e7026517f38f10ed760a90dbe2f)

stickies-v reviewed (approved):
re-ACK 1be04724a3ac061859118c09b90e2e15ea8d04b0

nit: these don't seem to actually have been addressed
> Addressed @stickies-v's https://github.com/bitcoin/bitcoin/pull/28186#discussion_r1283098892 and https://github.com/bitcoin/bitcoin/pull/28186#discussion_r1283280722, added some line breaks in constructor lists.

TheCharlatan commented:
Updated 1be04724a3ac061859118c09b90e2e15ea8d04b0 -> d8f1222ac50f089a0af29eaf8ce0555bad8366ef ([cleaveLeveldbHeaders_5](https://github.com/TheCharlatan/bitcoin/tree/cleaveLeveldbHeaders_5) -> [cleaveLeveldbHeaders_6](https://github.com/TheCharlatan/bitcoin/tree/cleaveLeveldbHeaders_6),  [compare](https://github.com/TheCharlatan/bitcoin/compare/cleaveLeveldbHeaders_5..cleaveLeveldbHeaders_6))

Sorry for the re-push @stickies-v, did not include two changes in my previous push by mistake that should have addresses your latest comments. Pushed now.

stickies-v commented:
re-ACK https://github.com/bitcoin/bitcoin/commit/d8f1222ac50f089a0af29eaf8ce0555bad8366ef

maflcko reviewed (approved):
ACK d8f1222ac50f089a0af29eaf8ce0555bad8366ef  🔠

<details><summary>Show signature</summary>

Signature:

```
untrusted comment: signature from minisign secret key on empty file; verify via: minisign -Vm "${path_to_any_empty_file}" -P RWTRmVTMeKV5noAMqVlsMugDDCyyTSbA3Re5AkUrhvLVln0tSaFWglOw -x "${path_to_this_whole_four_line_signature_blob}"
RUTRmVTMeKV5npGrKx1nqXCw5zeVHdtdYURB/KlyA/LMFgpNCs+SkW9a8N95d+U4AP1RJMi+krxU1A3Yux4bpwZNLvVBKy0wLgM=
trusted comment: ACK d8f1222ac50f089a0af29eaf8ce0555bad8366ef  🔠
9kmrmmglu0pLTAnlkFVLGRaV07BiskXJWhuBAXoJ8FtylkMHfRojj+S77g6nhGaMuZlTOzX7YETCJ4ZtX4yEAw==
```

</details>
